### PR TITLE
editor: Add support for customizable soft wrap indentation

### LINKED ISSUE #18687: Extra indent for soft-wrapped lines
### Check for existing issues

- [X] Completed

### Describe the feature

A soft-wrapped line is displayed with the same indentation as its beginning, which is hard to read. Other editors have options to add 1 or 2 indent sizes to these lines.

Proposal: add `extra_soft_wrap_indent` option.

Compare Zed:

Screenshot 2024-10-03 at 1 28 58 PM

and say Sublime Text:

Screenshot 2024-10-03 at 1 29 12 PM

### If applicable, add mockups / screenshots to help present your vision of the feature

_No response_

### PR DESCRIPTION
**Pull request by arusahni**
# Objective

- Soft wrapped lines currently inherit the indentation of their first line. This can be challenging for those coming from different editors, or with different expectations.
- Allow users to customize the soft wrap behavior to wrap with extra indent, or no indent at all.
- Fixes #18687

## Solution

- Introduce 3 soft-wrap indent configuration values (in addition to the default of `same`): `none`, `extra_one`, and `extra_two` (where they are multiples of the indent column width).
- Plumb indent configuration through to the various gpui and editor components that are concerned with wrapping.
- Update code action icon placement to accommodate `none`-level wrap indentation (preceded by a refactor of the function used to calculate its display row).
- Fixed an existing bug in `WrapSnapshot::soft_wrap_indent` where querying a 0-indent wrap returned `None` instead of `Some(0)`

## Testing

- Added tests to `wrap_map.rs` for all indentation levels.
- Updated the random wrap tests to include indent setting combinations. 
- Added tests for code action icon placement. 
- Validated locally by toggling between the various settings and observing how the wrap behaved.

## Self-Review Checklist:

- [x] I've reviewed my own diff for quality, security, and reliability
- [x] Unsafe blocks (if any) have justifying comments
- [x] The content adheres to Zed's UI standards ([UX/UI](https://github.com/zed-industries/zed/blob/main/CONTRIBUTING.md#uiux-checklist) and [icon](https://github.com/zed-industries/zed/blob/main/crates/icons/README.md) guidelines)
- [x] Tests cover the new/changed behavior
- [x] Performance impact has been considered and is acceptable

## Showcase

Changes are grouped by commit to ease review.

Screenshots below!

<details>
  <summary>Click to view showcase</summary>

**`same` indent** (current Zed behavior)
Screenshot from 2026-08-02 22-45-54

**`none` indent**
Screenshot from 2026-08-02 22-45-39

**`extra_one` indent**
Screenshot from 2026-08-02 22-46-06

**`extra_two` indent**
Screenshot from 2026-08-02 22-46-19

**the setting**
 Screenshot from 2026-08-02 22-42-19

</details>

---

Release Notes:

- Added a `soft_wrap_indent` setting (`"none"`, `"same"`, `"extra_one"`, `"extra_two"`) to configure indentation for soft-wrapped continuation lines.
- Fixed soft-wrap continuation line indentation calculation for unindented lines starting at column 0.

### CODE REVIEW

File: crates/settings/src/vscode_import.rs [hunk 1]
```diff
@@ -537,6 +537,7 @@ impl VsCodeSettings {
     fn default_language_settings_content(&self) -> LanguageSettingsContent {
         LanguageSettingsContent {
             allow_rewrap: None,
+            soft_wrap_indent: None,
```
Comment (SomeoneToIgnore) [inline crates/settings/src/vscode_import.rs]: This is not right: VSCode has `editor.wrappingIndent` that we need to import.

File: assets/settings/default.json [hunk 2]
```diff
@@ -1579,6 +1579,12 @@
   // 3. Soft wrap lines at the preferred line length or the editor width (whichever is smaller).
   //      "soft_wrap": "bounded",
   "soft_wrap": "none",
+  // How to indent soft-wrapped continuation lines.
+  //   "none"      - start at column 0
+  //   "same"      - match original line's indentation (default)
+  //   "extra_one" - same + 1 indent level
+  //   "extra_two" - same + 2 indent levels
+  "soft_wrap_indent": "same",
```
Comment (SomeoneToIgnore) [inline assets/settings/default.json:1627]: We seem to miss the all-settings.md entry here, let's add it too, so zed.dev/docs can showcase this feature.

File: crates/editor/src/editor.rs [hunk 3]
```diff
@@ -10171,6 +10175,7 @@ impl Editor {
                 self.clear_disabled_lsp_folding_ranges(window, cx);
                 self.refresh_document_symbols(None, cx);
                 self.refresh_outline_symbols_at_cursor(cx);
+                self.apply_soft_wrap_indent(cx);
```
Comment (SomeoneToIgnore) [inline crates/editor/src/editor.rs:10270]: The setting is per-language, but the push sites (here and line 2683) miss `multi_buffer::Event::LanguageChanged` (line 9983).

File: crates/editor/src/display_map/wrap_map.rs [hunk 4]
```diff
@@ -263,6 +265,22 @@ impl WrapMap {
...
+        }
+
+        self.soft_wrap_indent = indent;
+        self.rewrap(cx);
+        true
+    }
+
+    #[ztracing::instrument(skip_all)]
+    fn indent_adjustment(&self) -> gpui::IndentAdjustment {
+        indent_adjustment_for(self.soft_wrap_indent, self.snapshot.tab_snapshot.tab_size)
```
Comment (SomeoneToIgnore) [inline crates/editor/src/display_map/wrap_map.rs]: `extra_one`/`extra_two` go stale when `tab_size` changes.
A `tab_size` change emits one full-range tab edit (`tab_map.rs:49-61`), but `flush_edits` calls this method before `self.snapshot` is replaced, so the whole buffer is rewrapped with the old tab size; `set_soft_wrap_indent` can't save it because it early-returns when the enum is unchanged (line 270).

Later per-line edits use the new tab size, leaving mixed continuation indents across the buffer (error ≤ 2×|old−new| columns per line; visual only, self-heals as lines are edited).

File: crates/editor/src/editor_tests.rs [hunk 5]
```diff
@@ -45184,6 +45184,7 @@ async fn test_select_delimiters_expansion(cx: &mut TestAppContext) {
     cx.assert_editor_state("foo(«x, { a: 1 }ˇ»);");
 }
 
+
```
Comment (SomeoneToIgnore) [inline crates/editor/src/editor_tests.rs]: Seems not needed.

File: crates/editor/src/editor_tests.rs [hunk 6]
```diff
@@ -45778,3 +45779,87 @@ async fn test_scroll_range_hold_freezes_before_first_settled_frame(cx: &mut Test
...
+        .unwrap();
+
+    // The text wraps after 25 columns.
+    // Indent is 20, plus "12345" is 25. So it wraps before "6789".
+    // Buffer point at column 26 (the '7').
+    let buffer_point = Point::new(0, 26);
+
+    let display_row = snapshot.display_row_for_inline_code_action(buffer_point);
+
+    // With SoftWrapIndent::None, the wrapped line has 0 indent (0 < 14),
```
Comment (SomeoneToIgnore) [inline crates/editor/src/editor_tests.rs]: `(0 < 14)` — the limit is `INLINE_SLOT_CHAR_LIMIT = 4`, so this should read `(0 < 4)`.

File: crates/settings_content/src/language.rs [hunk 7]
```diff
@@ -506,6 +506,33 @@ impl<'de> Deserialize<'de> for ConfiguredLanguageServer {
...
+    /// Continuation lines start at column 0.
+    None,
+    /// Continuation lines match the original line's indentation.
+    #[default]
+    Same,
+    /// Continuation lines get 1 extra indent level beyond the original.
+    ExtraOne,
+    /// Continuation lines get 2 extra indent levels beyond the original.
+    ExtraTwo,
+}
```
Comment (SomeoneToIgnore) [inline crates/settings_content/src/language.rs:535]: Nit: missing blank line between the enum and the `LanguageSettingsContent` doc comment below.

File: crates/editor/src/display_map/wrap_map.rs [hunk 8]
```diff
@@ -886,7 +913,7 @@ impl WrapSnapshot {
         let (.., item) = self.transforms.find::<WrapPoint, _>(
             (),
             &WrapPoint::new(row + WrapRow(1), 0),
-            Bias::Right,
+            Bias::Left,
```
Comment (SomeoneToIgnore) [inline crates/editor/src/display_map/wrap_map.rs:919]: This is a fundamental change, can you provide more context what was that needed for?

**Review by SomeoneToIgnore — CHANGES_REQUESTED** [review]
Thank you, needs some more work.

File: crates/editor/src/display_map/wrap_map.rs (hunk 8 above)
Comment (arusahni) [reply crates/editor/src/display_map/wrap_map.rs:919]: This is a fix for what I believe is an existing issue that the zero-indent option revealed.

Prior to this feature, wrap transforms kept the source line's indentation, which is > 0 for any indented line. Their output endpoint would be `WrapPoint(row + 1, indent)` where `indent > 0`, which is greater than the search target of `WrapPoint(row + 1, 0)`. Both `Bias::Left` and `Bias::Right` found it identically, so the bug rarely arose.

When `indent == 0`, the search target and the wrap transform's endpoint are the same point. With `Bias::Right`, `find` treats an exact match as falling past that item to the next transform, which then returns `None` because it isn't a wrap. `Bias::Left` stops on the exact match and correctly returns `Some(0)`.

[This test demonstrates this edge case](https://github.com/zed-industries/zed/pull/63468/changes#diff-d635101d46f95cfa885dc200ad5f2a045b99d6dd59a0ff93184efb6c99015ca9R1588)

Comment (SomeoneToIgnore) [reply crates/editor/src/display_map/wrap_map.rs:919]: Ah, I see, we have never executed that in prod before, it seems?
Looks great, thank you for the explanation.

Comment (arusahni) [reply crates/editor/src/display_map/wrap_map.rs:919]: We may have, but this function currently appears to be unused on `main`. We used to use it for movement, but that changed in #2912 .

File: crates/settings/src/vscode_import.rs (hunk 1 above)
Comment (arusahni) [reply crates/settings/src/vscode_import.rs]: I have pushed an update that imports this. In doing so I realized that I was not familiar with VS Code's naming for these settings. Instead of `ExtraOne` and `ExtraTwo`, they use `Indent` and `DeepIndent`. Do you/Zed have a preference as to which nomenclature is used? My thinking with `ExtraN` was that additional indent multiples could be added in the future without running into t-shirt-sizing challenges (e.g. `ExtraDeeperIndent`).

File: assets/settings/default.json (hunk 2 above)
Comment (arusahni) [reply assets/settings/default.json:1627]: Good catch! Updated.

File: crates/editor/src/display_map/wrap_map.rs (hunk 4 above)
Comment (arusahni) [reply crates/editor/src/display_map/wrap_map.rs]: Good catch! I have updated `indent_adjustment` to accept a `tab_size` parameter so we can pass in the new snapshot's value at rewrap time.

File: crates/editor/src/editor_tests.rs (hunk 5 above)
Comment (arusahni) [reply crates/editor/src/editor_tests.rs]: Correct! Fixed via format.

File: crates/editor/src/editor_tests.rs (hunk 6 above)
Comment (arusahni) [reply crates/editor/src/editor_tests.rs]: Ah yeah, stale comment from an earlier version. :see_no_evil:

File: crates/settings_content/src/language.rs (hunk 7 above)
Comment (arusahni) [reply crates/settings_content/src/language.rs:535]: Fixed via format.

File: crates/editor/src/editor.rs (hunk 3 above)
Comment (arusahni) [reply crates/editor/src/editor.rs:10270]: I have addressed this by calling `apply_soft_wrap_indent(cx)` in the `multi_buffer::Event::LanguageChanged` branch (and added a test to verify).

**Comment by arusahni** [comment]
@SomeoneToIgnore thanks for the review! I believe I have addressed all your comments. There is an open question I have around the names for the setting values, but thankfully it shouldn't be a huge lift if we decide to go with different nomenclature.

File: crates/editor/src/editor.rs (hunk 3 above)
Comment (SomeoneToIgnore) [reply crates/editor/src/editor.rs:10270]: Moving an open file between directories with different `soft_wrap_indent` settings misses this apply call.
The buffer refreshes its resolved settings, but the editor’s [`SettingsChanged` handler](https://github.com/zed-industries/zed/blob/8f38a40504b8f4694b61b5906faafbd9a0e31892/crates/editor/src/editor.rs#L10077-L10088) only updates `applicable_language_settings`.
A same-language move does not produce the `LanguageChanged` event that would repair it.

For a four-space-indented line moved from `"same"` into `"none"`, continuation lines retain four columns instead of zero.
Please apply the indent on the buffer-level settings event too.

File: crates/editor/src/config.rs [hunk 9]
```diff
@@ -257,6 +257,22 @@ impl Editor {
         wrap_guides
     }
 
+    pub(super) fn soft_wrap_indent(&self, cx: &App) -> SoftWrapIndent {
+        self.buffer.read(cx).language_settings(cx).soft_wrap_indent
+    }
+
+    pub(super) fn apply_soft_wrap_indent(&mut self, cx: &mut Context<Self>) {
```
Comment (SomeoneToIgnore) [inline crates/editor/src/config.rs:264]: This captures global defaults when the multibuffer is empty, but adding its first excerpt can change the effective language settings.
Project search creates its results editor before inserting excerpts from result buffers.
Inserting an already-loaded buffer emits `Edited` and `BufferRangesUpdated`, but [neither handler reapplies the indent](https://github.com/zed-industries/zed/blob/8f38a40504b8f4694b61b5906faafbd9a0e31892/crates/editor/src/editor.rs#L9924-L10006).

With global `"same"` and the first result buffer configured as `"none"`, a four-space-indented line retains four continuation columns in search results while its source editor uses zero.
Please refresh this state when the first excerpt changes.

File: crates/editor/src/editor.rs [hunk 10]
```diff
@@ -12018,6 +12024,139 @@ impl EditorSnapshot {
...
+                    true
+                } else if row_candidate == buffer_point.row {
+                    // If the initial line indentation is less than the slot limit,
+                    // check if the cursor is on a soft-wrapped continuation row whose
+                    // wrap indentation provides enough space.
+                    let display_row = self
+                        .display_snapshot
+                        .point_to_display_point(buffer_point, text::Bias::Left)
+                        .row();
+                    if display_row.0 > 0 {
```
Comment (SomeoneToIgnore) [inline crates/editor/src/editor.rs:12098]: `display_row - 1` can be a code-lens or excerpt-header block rather than the preceding text row.
An `Above` block maps to the following text row, so this lookup reads that row’s **outgoing** wrap indent and mistakes it for whitespace before the cursor’s row.

For `12345678901234567890123456789`, wrapping at column 25 with four continuation spaces and a one-row block above, cursor column 5 incorrectly returns `Some(DisplayRow(1))`.
That first text row has no leading whitespace.
Please prove that the cursor is on a continuation before using this exception.

File: crates/editor/src/editor_tests.rs [hunk 11]
```diff
@@ -45937,3 +45937,124 @@ async fn test_scroll_range_hold_freezes_before_first_settled_frame(cx: &mut Test
...
+    let buffer_point = Point::new(0, 26);
+    let display_row = snapshot.display_row_for_inline_code_action(buffer_point);
+
+    // Physical line 0 has 2 spaces indent (< 4), but wrapped DisplayRow(1) has 6 spaces (>= 4).
+    // The code action helper should stay on physical line 0 and render on DisplayRow(1).
+    assert_eq!(display_row, Some(DisplayRow(1)));
+}
+
+#[gpui::test]
+async fn test_soft_wrap_indent_updated_on_language_changed(cx: &mut gpui::TestAppContext) {
```
Comment (SomeoneToIgnore) [inline crates/editor/src/editor_tests.rs:46026]: This does not seem to test the settings propagation?

Both assertions read the settings getter, not the display map, so removing the new apply call does not affect their dependencies.

Use nonempty wrapped text and assert its continuation indentation before and after changing language.

File: crates/editor/src/editor.rs [hunk 12]
```diff
@@ -12018,6 +12024,139 @@ impl EditorSnapshot {
             .language_at(position)
     }
 
+    pub fn display_row_for_inline_code_action(&self, buffer_point: Point) -> Option<DisplayRow> {
+        // do not show code action for folded line
```
Comment (SomeoneToIgnore) [inline crates/editor/src/editor.rs:12028]: LLM-esque comments explaining the few lines verbatim are not needed here and in other places — please remove them.

**Review by SomeoneToIgnore — CHANGES_REQUESTED** [review]
Getting closer, thank you.

File: crates/editor/src/editor.rs (hunk 12 above)
Comment (arusahni) [reply crates/editor/src/editor.rs:12028]: Got it. These comments were already in the code (I refactored the block into a function). I'll remove them.

**Comment by arusahni** [comment]
Thanks for the second review! I won't be able to address these comments until next week because I'll be at RustConf.

File: crates/editor/src/editor.rs (hunk 12 above)
Comment (arusahni) [reply crates/editor/src/editor.rs]: These comments have been removed.

File: crates/editor/src/editor.rs (hunk 3 above)
Comment (arusahni) [reply crates/editor/src/editor.rs:10397]: I appreciate you catching this. I have made the change, and added a test to validate this works as expected when moving a file.

File: crates/editor/src/editor_tests.rs (hunk 11 above)
Comment (arusahni) [reply crates/editor/src/editor_tests.rs:49677]: Yeah, oversight on my part. I have made this test actually test something :sweat_smile:

File: crates/editor/src/editor.rs (hunk 10 above)
Comment (arusahni) [reply crates/editor/src/editor.rs]: Thanks for flagging this - I'm learning the editor as I go. I have implemented this, and added a test for validation.

File: crates/editor/src/config.rs (hunk 9 above)
Comment (arusahni) [reply crates/editor/src/config.rs:264]: I have added recalculations for `BufferRangesUpdated` and `BuffersRemoved`, and a test to validate this behavior.

**Comment by arusahni** [comment]
I've addressed all additional feedback as discrete commits and rebased this over `main`, addressing a merge conflict in `editor_tests.rs`.

File: crates/gpui/src/text_system/line_wrapper.rs [hunk 13]
```diff
@@ -107,9 +120,15 @@ impl LineWrapper {
...
-                            Self::MAX_INDENT.min((first_non_whitespace_ix - last_wrap_ix) as u32),
-                        );
+                        let base_indent =
+                            Self::MAX_INDENT.min((first_non_whitespace_ix - last_wrap_ix) as u32);
+                        indent = Some(match indent_adjustment {
+                            IndentAdjustment::NoIndent => 0,
+                            IndentAdjustment::SameIndent => base_indent,
+                            IndentAdjustment::ExtraColumns(extra) => {
+                                Self::MAX_INDENT.min(base_indent + extra)
+                            }
```
Comment (SomeoneToIgnore) [inline crates/gpui/src/text_system/line_wrapper.rs:130]: `SameIndent` is self-bounding: `base_indent` is measured inside a row that fit, so it never exceeds the wrap column count.
`ExtraColumns` adds up to `2 × tab_size` on top with only the `MAX_INDENT` clamp, so `indent` can exceed the wrap width.
After that, `width = item_width + indent × space_width > wrap_width` holds for every following character, and the line degrades to one character per row, each row wider than the wrap width.

Repro (`EditorTestContext`, `soft_wrap: bounded`, `preferred_line_length: 10`, `extra_two`, tab size 4, text `"      ab cd ef gh"`):

```
|      ab |
|              c|
|              d|
|               |
|              e|
|              f|
|               |
|              g|
|              h|
```

9 rows of 15 columns in a 10-column wrap; `same` gives 3 rows of ≤ 10 columns.
The overflow is bounded by `2 × tab_size` columns, with one row per remaining character, so a 200-character line becomes ~200 rows.
This is reachable with `extra_two` for any line whose indent is within 8 columns of the pane width, e.g. deeply nested code in a narrow split.

VS Code guards exactly this in `computeWrappedTextIndentLength`: `if (wrappedTextIndentLength + columnsForFullWidthChar > firstLineBreakColumn) wrappedTextIndentLength = 0`.
Please add the same guard here (fall back to `0`, or to `base_indent`) and cover it with a unit test in this file.

File: crates/editor/src/config.rs (hunk 9 above)
Comment (SomeoneToIgnore) [inline crates/editor/src/config.rs:264]: The existing precedent for a settings-driven wrap parameter is `soft_wrap`: `element.rs` calls `editor.soft_wrap_mode(cx)` + `set_wrap_width` on every prepaint (lines 8579-8589 and 11286-11290), and `set_wrap_width` early-returns when unchanged.
`set_soft_wrap_indent` already early-returns the same way.
Reading `self.soft_wrap_indent(cx)` inside `Editor::set_wrap_width` below and forwarding it to whichever map is picked there gives the indent the same freshness as `soft_wrap`/`preferred_line_length`, and removes the `Editor::new`, `is_minimap`, and five event-handler call sites.

It also closes the remaining gap: an empty multibuffer (project search before the first result) whose global `soft_wrap_indent` changes emits no buffer `SettingsChanged`, and `fetch_applicable_language_settings` compares two empty maps, so `placeholder_display_map` stays stale until an excerpt arrives.

The new tests keep working unmodified: in test mode `App::flush_effects` draws dirty windows (`gpui/src/app.rs:1817-1830`), so the poll runs before the next `editor.update`.
Cost is one extra `MultiBuffer::language_settings` per frame, which `soft_wrap_mode` already pays on the same frame.

File: crates/settings/src/vscode_import.rs [hunk 14]
```diff
@@ -552,6 +552,13 @@ impl VsCodeSettings {
     fn default_language_settings_content(&self) -> LanguageSettingsContent {
         LanguageSettingsContent {
             allow_rewrap: None,
+            soft_wrap_indent: self.read_enum("editor.wrappingIndent", |s| match s {
+                "none" => Some(SoftWrapIndent::None),
+                "same" => Some(SoftWrapIndent::Same),
+                "indent" => Some(SoftWrapIndent::ExtraOne),
```
Comment (SomeoneToIgnore) [inline crates/settings/src/vscode_import.rs]: On the naming question: let's keep `none`/`same`/`extra_one`/`extra_two`.
`"soft_wrap_indent": "indent"` reads badly, and this import already maps the VS Code names onto ours.

FYI only, no change needed: VS Code adds tab stops rather than flat `tab_size` columns, so base indent 6 with tab size 4 gives 8 there and 10 here.

File: crates/editor/src/display_map/wrap_map.rs [hunk 15]
```diff
@@ -1473,6 +1515,185 @@ mod tests {
...
+        // Change tab_size to 2
+        let (tabs_snapshot, tab_edits) =
+            tab_map.sync(fold_snapshot, Vec::new(), 2.try_into().unwrap());
+        let (wrap_snapshot, _) =
+            wrap_map.update(cx, |map, cx| map.sync(tabs_snapshot, tab_edits, cx));
+        assert_eq!(wrap_snapshot.text(), "    let x = \n      1;\n");
+    }
+
+    #[gpui::test]
+    async fn test_soft_wrap_indent_returns_some_for_zero_indent(cx: &mut gpui::TestAppContext) {
```
Comment (SomeoneToIgnore) [inline crates/editor/src/display_map/wrap_map.rs:1635]: The `None` case in `test_soft_wrap_indent` above already produces a zero-indent wrap.
One `assert_eq!(wrap_snapshot.soft_wrap_indent(WrapRow(1)), Some(0))` there covers this, so this test and its comment block can go.
The same 30 lines of setup are also copied into `test_soft_wrap_indent_updates_on_tab_size_change`; please share a helper if it stays.

File: crates/editor/src/display_map/wrap_map.rs [hunk 16]
```diff
@@ -263,6 +265,22 @@ impl WrapMap {
...
+            return false;
+        }
+
+        self.soft_wrap_indent = indent;
+        self.rewrap(cx);
+        true
+    }
+
+    #[ztracing::instrument(skip_all)]
+    fn indent_adjustment(&self, tab_size: std::num::NonZeroU32) -> gpui::IndentAdjustment {
```
Comment (SomeoneToIgnore) [inline crates/editor/src/display_map/wrap_map.rs:280]: Nit: `gpui::IndentAdjustment` and `std::num::NonZeroU32` are spelled out inline throughout this file, please import them.
The `ztracing` instrument on a one-line match is noise.

**Review by SomeoneToIgnore — CHANGES_REQUESTED** [review]
Seems that we need to deal with `line_wrapper` code better, also added a few more notes.
